:memo: Rework Custom certificates (#2771)
platformsh/platformsh-docs: docs/src/domains/troubleshoot.md

If it isn't, try the following steps:

- Your DNS server might not be correctly configured or the old DNS records are still cached.
  Try removing your local DNS cache.
- Set your computer's DNS server to Google Public DNS (`8.8.8.8` and `8.8.4.4`)
  to see if the issue is with the DNS server you are using.
- Run `ping www.{{< variable "YOUR_DOMAIN" >}}`.
  If the result is different from what you got from running `host www.{{< variable "YOUR_DOMAIN" >}}`,
  you might need to remove your test settings.

## Verify SSL

For troubleshooting, feel free to use [the certificate checker tool](https://certcheck.pltfrm.sh/).
This tool assists in finding out where your domain is pointing to and provides some generic guidance.
It also assists when a CDN (Fastly, Cloudflare, ...) is used.
It's good practice to check both the apex and the `www` domain to ensure that both point to your project.

> For more in-depth investigations, on the command line using macOS, Linux, or the Windows Subsystem for Linux, run the following command:

Author: Much much better! It's clearer, and shell is the perfect fit there.

```bash
curl -I -v  https://www.{{< variable "YOUR_DOMAIN" >}}
```

The response should be long. Look for error messages.
They're usually explicit enough.
Often the problem is caused by a mismatch between the certificate and the domain name or an expired [custom certificate](steps/tls.md).

> ### Error provisioning certificates

Author: :clipboard: That part should fit a bit better to the rest of the page and it's hopefully clearer, but might need more rework to be shorter and clearer.

```

The renewal may fail because of the 64-character limit Let's Encrypt places on URLs.
If you have a branch with a long name, the environment URL is over this limit and the certificate is rejected.
Shortening the branch name to fewer than 20 characters should solve the issue.

Generated URLs for environments have the following pattern:

```bash
{{<variable "ENVIRONMENT" >}}-{{<variable "PROJECT_ID" >}}.{{<variable "REGION" >}}.platformsh.site
```

If you have a [default domain](../define-routes/_index.md#default) set up, the generated URL has the following pattern:

```bash
{{<variable "YOUR_DOMAIN" >}}.{{<variable "ENVIRONMENT" >}}-{{<variable "PROJECT_ID" >}}.{{<variable "REGION" >}}.platformsh.site
```

The generated URLs consist of:

- `{{<variable "YOUR_DOMAIN" >}}` = the amount of characters your domain has
- `{{<variable "ENVIRONMENT" >}}` = `{{<variable "BRANCH_NAME" >}}` + 7 character hash
- `{{<variable "PROJECT_ID" >}}` = 13 characters
- `{{<variable "REGION" >}}` = 2 to 4 characters, depending on the region
- `platformsh.site` = 15 characters
- extra characters like `.` and `-` = 4 to 5 characters, depending on if you have a default domain

> This leaves you with 21 to 23 characters for your branch name (`{{<variable "BRANCH_NAME" >}}`) without going over the 64-character limit,
> depending on the region.
> Since this pattern for generated URLs should remain similar even if it may change slightly,
> your branch names should be no more than 20 characters.

Reviewer: 🛠️ 

```suggestion
This leaves you with 21 to 23 characters for your branch name (`{{<variable "BRANCH_NAME" >}}`) without exceeding the 64-character limit,
depending on the region.
To ensure your renewals succeed,  
keep your branch names under 20 characters.
```

### DNS Challenge

To provide a valid SSL-certificate,
Let's Encrypt needs to ensure that the requester is entitled to receive the SSL-certificate it asked for

> (usually through the presence of a specific token on the DNS zone of that domain).

Reviewer: 🛠️ I think you can split this long sentence into two to make it easier to read/ [avoid the use of parentheses](https://developers.google.com/style/parentheses).

This ownership verification is achieved through the so-called [_Challenge_ step](https://letsencrypt.org/docs/challenge-types/).

The certificate request is generated based on your [routes definition](../define-routes/_index.md).
If you want your site to be available with `example.com` and its `www.example.com` subdomain, make sure both are defined in your routes.

> Platform.sh checks that all the routes you defined are pointing to your project, if that's not the case, the verification fails, which results in the following error-message:

Reviewer: 🛠️ To shorten sentences.

```suggestion
Platform.sh checks that all the routes you defined are pointing to your project.
If not, the verification fails and you get the following error-message:
```

```text
Couldn't complete challenge [HTTP01: pending | DNS01: pending | TLSALPN01: pending]
```

> For the DNS challenge to complete, domains, and subdomains must point directly to your Platform.sh project (unless using a [CDN](../domains/cdn/_index.md)).

Reviewer: 🛠️ 

```suggestion
For the DNS challenge to complete, 
domains and subdomains must point directly to your Platform.sh project, 
unless you use a [CDN](../domains/cdn/_index.md)).
```

Otherwise, you see the following error:

```text
  E: Error validating domain www.example.com: Couldn't complete challenge [HTTP01: pending | DNS01: pending | TLSALPN01: pending]
  Unable to validate domains www.example.com, will retry in the background.
```

or

```text
  W: Failed to verify the challenge at the gateway for the domain 'www.some-example.example.com'
  E: Error validating domain www.example.com: Couldn't complete challenge [HTTP01: There was a problem with a DNS query during identifier validation]
```

Make sure that the [apex domain](../other/glossary.md#apex-domain) and its `www` subdomain are both pointing to your project.

> Note that it can take anywhere from 15 minutes to 72 hours for DNS changes to be taken into account.

Reviewer: 🛠️ To shorten sentence.

```suggestion
Note that it can take up to 72 hours for DNS changes to be effective.
```

> See the [step-by-step guide](../domains/steps/_index.md) for more information.

Reviewer: 💅 

```suggestion
For more information, see how to [set up a custom domain](../domains/steps/_index.md).
```

> If you have waited a couple of hours, properly configured the subdomain, and are still seeing an error of this type,
> [redeploying](../development/troubleshoot.md#force-a-redeploy) the impacted environment usually solves the issue.

Reviewer: 🛠️  To make it more straightforward/less wordy.

```suggestion
If the changes take longer than expected,
[redeploy](../development/troubleshoot.md#force-a-redeploy) the impacted environment.
```

Also make sure that no conflicting DNS records exist for the domain.
For example, a conflicting AAAA (IPv6) DNS record usually results in a `[HTTP01: The client lacks sufficient authorization]` error.

> If the certificate generation issue persists,
> you could also verify if an outage is currently ongoing on [with Let's Encrypt](https://letsencrypt.status.io/).
> If that isn't the case, [contact support](../overview/get-support.md).

Reviewer: 🛠️ To make it more straightforward.

```suggestion
If the certificate generation issue persists,
check if an outage is ongoing [with Let's Encrypt](https://letsencrypt.status.io/).
If not, [contact Support](../overview/get-support.md).
```

Generated URLs for environments have the following pattern:

```bash
{{<variable "ENVIRONMENT" >}}-{{<variable "PROJECT_ID" >}}.{{<variable "REGION" >}}.platformsh.site
```

If you have a [default domain](../define-routes/_index.md#default) set up, the generated URL has the following pattern:

```bash
{{<variable "YOUR_DOMAIN" >}}.{{<variable "ENVIRONMENT" >}}-{{<variable "PROJECT_ID" >}}.{{<variable "REGION" >}}.platformsh.site
```

The generated URLs consist of:

- `{{<variable "YOUR_DOMAIN" >}}` = the amount of characters your domain has
- `{{<variable "ENVIRONMENT" >}}` = `{{<variable "BRANCH_NAME" >}}` + 7 character hash
- `{{<variable "PROJECT_ID" >}}` = 13 characters
- `{{<variable "REGION" >}}` = 2 to 4 characters, depending on the region
- `platformsh.site` = 15 characters
- extra characters like `.` and `-` = 4 to 5 characters, depending on if you have a default domain

This leaves you with 21 to 23 characters for your branch name (`{{<variable "BRANCH_NAME" >}}`) without exceeding the 64-character limit,
depending on the region.
To ensure your renewals succeed, 
keep your branch names under 20 characters.

> ### DNS Challenge

Reviewer: ❓ I think Tim had a point when he said the phrase DNS challenge could be confused with the [DNS-01 challenge](https://letsencrypt.org/docs/challenge-types/#dns-01-challenge), which seems to have been replaced at PSH by the [HTTP-01 challenge](https://letsencrypt.org/docs/challenge-types/#http-01-challenge) => to be confirmed cf. our previous conversation. In any event, I think his suggestion to use "Let's Encrypt challenge" to avoid confusion is a good one, don't you?

Author: I think it's maybe not a good idea to make it too Let's Encrypt centric, to make it fit better with the CDN (and fastly) altough it's a very good point so I tried to adapt accordingly.
Please let me know if that'd work for you.

To provide a valid SSL-certificate,
Let's Encrypt checks that the requester is entitled to receive the requested SSL certificate.
This check is known as the [_Challenge_ step](https://letsencrypt.org/docs/challenge-types/).

The certificate request is generated based on your [routes definition](../define-routes/_index.md).
If you want your site to be available with `example.com` and its `www.example.com` subdomain, make sure both are defined in your routes.

> {{< codetabs >}}

Author: I hope it's better now. I think the split does make sense to reuse content where possible.
